crates/aptos-crypto/src/bls12381.rs

```diff
@@ -0,0 +1,263 @@
+        assert!(signature_wrong
+            .verify_arbitrary_msg(message, &key_pair_wrong.public_key)
+            .is_ok());
+
+        // sig on message under keypair should NOT verify under keypair_wrong
+        assert!(signature
+            .verify_arbitrary_msg(message, &key_pair_wrong.public_key)
+            .is_err());
+
+        // sig on message under keypair should NOT verify on message_wrong under key_pair
+        assert!(signature
+            .verify_arbitrary_msg(message_wrong, &key_pair.public_key)
+            .is_err());
+
+        // sig on message under keypair_wrong should NOT verify under key_pair
+        assert!(signature_wrong
+            .verify_arbitrary_msg(message, &key_pair.public_key)
+            .is_err());
+    }
+
+    // NOTE(Alin): The whole point of a multisig scheme is to avoid batch verifying a bunch of
+    // signatures by aggregating them into a multisignature. So we don't need this.
+    //
+    // #[test]
+    // fn bls12381_batch_verify() {
+    //     let message = TestAptosCrypto("lalala".into());
+    //     let mut rng = OsRng;
+    //     let mut key_pairs = vec![];
+    //     for _ in 0..1000 {
+    //         key_pairs.push(KeyPair::<BLS12381PrivateKey, BLS12381PublicKey>::generate(
+    //             &mut rng,
+    //         ));
+    //     }
+    //     let mut keys_and_signatures = vec![];
+    //     for keys in key_pairs {
+    //         let signature = keys.private_key.sign(&message);
+    //         keys_and_signatures.push((keys.public_key.clone(), signature));
+    //     }
+    //     assert!(BLS12381Signature::batch_verify(&message, keys_and_signatures).is_ok());
+    // }
+
+    #[test]
+    fn bls12381_pop_verify() {
+        let mut rng = OsRng;
+
+        let keypair1 = KeyPair::<BLS12381PrivateKey, BLS12381PublicKey>::generate(
+            &mut rng,
+        );
+
+        let keypair2 = KeyPair::<BLS12381PrivateKey, BLS12381PublicKey>::generate(
+            &mut rng,
+        );
+
+        // Valid PoP for SK 1
+        let pop1 = BLS12381ProofOfPossession::create(&keypair1.private_key, &keypair1.public_key);
+        // Valid PoP for SK 2
+        let pop2 = BLS12381ProofOfPossession::create(&keypair2.private_key, &keypair2.public_key);
+        // Invalid PoP for SK 2
+        let pop_bad = BLS12381ProofOfPossession::create(&keypair1.private_key, &keypair2.public_key);
+
+        // PoP for SK i should verify for PK i
+        assert!(pop1.verify(&keypair1.public_key).is_ok());
+        assert!(pop2.verify(&keypair2.public_key).is_ok());
+
+        // PoP for SK 1 should not verify for PK 2
+        assert!(pop1.verify(&keypair2.public_key).is_err());
+        // Pop for SK 2 should not verify for PK 1
+        assert!(pop2.verify(&keypair1.public_key).is_err());
+        // Invalid PoP for SK 2 should not verify
+        assert!(pop_bad.verify(&keypair2.public_key).is_err());
+    }
+
+    #[test]
+    fn bls12381_multisig_verify() {
```

Comment:
since this is the most concerned use case, I'd hope for more test scenarios. e.g. multisig with bad signature shares (signed with wrong message, signed with wrong key etc) aggregated pubkey with a few bad keys etc.